File: server/src/main/java/dev/slimevr/autobone/errors/PositionError.java
package dev.slimevr.autobone.errors;

import java.util.List;

import com.jme3.math.FastMath;

import dev.slimevr.autobone.AutoBoneTrainingStep;
import dev.slimevr.poserecorder.PoseFrameTracker;
import dev.slimevr.poserecorder.TrackerFrame;
import dev.slimevr.poserecorder.TrackerFrameData;
import dev.slimevr.tracking.processor.skeleton.HumanSkeleton;
import dev.slimevr.tracking.trackers.ComputedTracker;


// The distance of any points to the corresponding absolute position
public class PositionError implements IAutoBoneError {
	@Override
	public float getStepError(AutoBoneTrainingStep trainingStep) throws AutoBoneException {
		List<PoseFrameTracker> trackers = trainingStep.getTrainingFrames().getTrackers();
		return (getPositionError(
			trackers,
			trainingStep.getCursor1(),
			trainingStep.getHumanPoseManager1().getSkeleton()
		)
			+ getPositionError(
				trackers,
				trainingStep.getCursor2(),
				trainingStep.getHumanPoseManager2().getSkeleton()
			))
			/ 2f;
	}

	public static float getPositionError(
		List<PoseFrameTracker> trackers,
		int cursor,
		HumanSkeleton skeleton
	) {
		float offset = 0f;
		int offsetCount = 0;

		for (PoseFrameTracker tracker : trackers) {
			TrackerFrame trackerFrame = tracker.safeGetFrame(cursor);
			if (
				trackerFrame == null
					|| !trackerFrame.hasData(TrackerFrameData.POSITION)
					|| !trackerFrame.hasData(TrackerFrameData.DESIGNATION)
					|| trackerFrame.designation.trackerRole.isEmpty()
			) {
				continue;
			}

			ComputedTracker computedTracker = skeleton
				.getComputedTracker(trackerFrame.designation.trackerRole.get());
			if (computedTracker != null) {
				offset += FastMath.abs(computedTracker.position.distance(trackerFrame.position));
				offsetCount++;
			}
		}

		return offsetCount > 0 ? offset / offsetCount : 0f;
	}
}

File: server/src/main/java/dev/slimevr/autobone/errors/PositionError.kt
package dev.slimevr.autobone.errors

import dev.slimevr.autobone.AutoBoneTrainingStep
import dev.slimevr.poseframeformat.trackerdata.TrackerFrames
import dev.slimevr.tracking.processor.skeleton.HumanSkeleton

// The distance of any points to the corresponding absolute position
class PositionError : IAutoBoneError {
	@Throws(AutoBoneException::class)
	override fun getStepError(trainingStep: AutoBoneTrainingStep): Float {
		val trackers = trainingStep.trainingFrames.frameHolders
		return (
			(
				getPositionError(
					trackers,
					trainingStep.cursor1,
					trainingStep.humanPoseManager1.skeleton
				) +
					getPositionError(
						trackers,
						trainingStep.cursor2,
						trainingStep.humanPoseManager2.skeleton
					)
				) /
				2f
			)
	}

	companion object {
		fun getPositionError(
			trackers: List<TrackerFrames>,
			cursor: Int,
			skeleton: HumanSkeleton,
		): Float {
			var offset = 0f
			var offsetCount = 0
			for (tracker in trackers) {
				val trackerFrame = tracker.tryGetFrame(cursor) ?: continue
				val position = trackerFrame.tryGetPosition() ?: continue
				val trackerRole = trackerFrame.tryGetTrackerPosition()?.trackerRole ?: continue

				val computedTracker = skeleton.getComputedTracker(trackerRole) ?: continue

				offset += (position - computedTracker.position).len()
				offsetCount++
			}
			return if (offsetCount > 0) offset / offsetCount else 0f
		}
	}
}
